### PR TITLE
Simplify observers declaration with functools.partial

### PR DESCRIPTION
**Pull request by dzhulgakov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26492 Simplify observers declaration with functools.partial**

Previous definition of observers was quite clumsy - with things like `default_observer()()`. This PR strips a way a lot of craft and allows to pass just class names directly. In order to override default arguments either `functools.partial` can be used or convenient wrapper `MyObserver.with_args(x=1)` is provided.

Also rename `QConfig_dynamic` to `QConfigDynamic` because it violates the naming convention.

Differential Revision: [D17521265](https://our.internmc.facebook.com/intern/diff/D17521265)

### CODE REVIEW

File: torch/quantization/QConfig.py [hunk 1]
```diff
@@ -2,19 +2,60 @@
...
-QConfig = namedtuple('QConfig',
-                     ['activation', 'weight'])
+class QConfig(namedtuple('QConfig', ['activation', 'weight'])):
+    """
+    Describes how to quantize a layer or a part of the network by providing
+    settings (observer classes) for activations and weights respectively.
 
-default_qconfig = QConfig(activation=default_observer(),
-                          weight=default_weight_observer())
+    Note that QConfig needs to contain observer **classes** (like MinMaxObserver), not the concrete instances.
```
Comment (jerryzh168) [inline torch/quantization/QConfig.py]: I think it's either **classes** or **functools.partial object**?

File: test/common_quantization.py [hunk 2]
```diff
@@ -279,8 +280,8 @@ def __init__(self):
             'dtype': torch.quint8,
             'qscheme': torch.per_tensor_affine
         }
-        custom_qconfig = QConfig(activation=default_observer(**custom_options),
-                                 weight=default_weight_observer())
+        custom_qconfig = QConfig(activation=partial(default_observer, **custom_options),
```
Comment (jerryzh168) [inline test/common_quantization.py]: I think we can do `with_args` here?

File: torch/quantization/QConfig.py (hunk 1 above)
Comment (jerryzh168) [reply torch/quantization/QConfig.py]: or we can say, it's something that will return an instance when called. sounds a bit confusing though.

**Review by jerryzh168 — APPROVED** [review]
Looks good Thanks

File: test/common_quantization.py (hunk 2 above)
Comment (dzhulgakov) [reply test/common_quantization.py]: oh, yeah, missed this one, thanks :)

**Pull request closed without merging**